From df7eeed31e1562397cbf814cc91585ee6df16ebc Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Tue, 21 May 2024 23:22:38 -0700
Subject: [PATCH] ath12k: Handle duplicate tx_desc from wbm2sw status ring

When a duplicate tx_desc is obtained from the wbm2sw status ring
in the tx completion path, it ends up getting added to the
free_list again, which leads to a double-free scenario.
Avoid this by by checking the in_use bit and capture the stats.

Signed-off-by: Nithyanantham Paramasivam <quic_nithp@quicinc.com>
Signed-off-by: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>
Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h    |  1 +
 drivers/net/wireless/ath/ath12k/debugfs.c |  7 +++++++
 drivers/net/wireless/ath/ath12k/dp_tx.c   | 15 +++++++++++----
 3 files changed, 19 insertions(+), 4 deletions(-)

Index: backports-6.5-rc3/drivers/net/wireless/ath/ath12k/core.h
===================================================================
--- backports-6.5-rc3.orig/drivers/net/wireless/ath/ath12k/core.h
+++ backports-6.5-rc3/drivers/net/wireless/ath/ath12k/core.h
@@ -1286,6 +1286,7 @@ struct ath12k_soc_dp_stats {
 	u32 bond_tx_mcast_enqueued[MAX_TCL_RING];
 	u32 bond_tx_ucast_dropped[MAX_TCL_RING];
 	u32 bond_tx_mcast_dropped[MAX_TCL_RING];
+	u32 duplicate_tx_desc[MAX_TCL_RING];
 	/* Number of frames successfully converted to ucast from mcast */
 	u32 bond_mcast2ucast_converted[MAX_TCL_RING];
 	/* Number of mcast frames tried for ucast conversion*/
Index: backports-6.5-rc3/drivers/net/wireless/ath/ath12k/debugfs.c
===================================================================
--- backports-6.5-rc3.orig/drivers/net/wireless/ath/ath12k/debugfs.c
+++ backports-6.5-rc3/drivers/net/wireless/ath/ath12k/debugfs.c
@@ -1740,6 +1740,13 @@ static ssize_t ath12k_debugfs_dump_soc_d
 			soc_stats->null_tx_complete[3]);
 
 	len += scnprintf(buf + len, size - len,
+			"\nduplicate tx desc: 0:%u 1:%u 2:%u 3:%u\n",
+			soc_stats->duplicate_tx_desc[0],
+			soc_stats->duplicate_tx_desc[1],
+			soc_stats->duplicate_tx_desc[2],
+			soc_stats->duplicate_tx_desc[3]);
+
+	len += scnprintf(buf + len, size - len,
 			"\ntx_completed: 0:%u 1:%u 2:%u 3:%u\n",
 			soc_stats->tx_completed[0],
 			soc_stats->tx_completed[1],
Index: backports-6.5-rc3/drivers/net/wireless/ath/ath12k/dp_tx.c
===================================================================
--- backports-6.5-rc3.orig/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ backports-6.5-rc3/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -1584,17 +1584,18 @@ int ath12k_dp_tx_completion_handler(stru
 			continue;
 		}
 
+		/* Handle duplicate tx_desc from wbm2sw status ring */
+		if (unlikely(!tx_desc->in_use)) {
+			ab->soc_stats.duplicate_tx_desc[tx_ring->tcl_data_ring_id]++;
+			continue;
+		}
+
 		msdu = tx_desc->skb;
 		mac_id = tx_desc->mac_id;
 		recycler_fast_xmit = tx_desc->recycler_fast_xmit;
 		skb_ext_desc = tx_desc->skb_ext_desc;
 		timestamp = tx_desc->timestamp;
 
-		/* Release descriptor as soon as extracting necessary info
-		 * to reduce contention
-		 */
-		tx_desc->in_use = false;
-		list_add_tail(&tx_desc->list, &desc_free_list);
 
 		if (unlikely(!msdu)) {
 			ab->soc_stats.null_tx_complete[tx_ring->tcl_data_ring_id]++;
@@ -1608,6 +1609,12 @@ int ath12k_dp_tx_completion_handler(stru
 			continue;
 		}
 
+		/* Release descriptor as soon as extracting necessary info
+		 * to reduce contention
+		 */
+		tx_desc->in_use = false;
+		list_add_tail(&tx_desc->list, &desc_free_list);
+
 		prefetch(msdu);
 		prefetch( (u8 *) msdu + 64);
 
